db/db_impl_compaction_flush.cc

```diff
@@ -200,6 +202,172 @@ Status DBImpl::FlushMemTableToOutputFile(
+    const autovector<uint64_t>& flush_memtable_ids, bool* made_progress,
+    JobContext* job_context, LogBuffer* log_buffer) {
+  mutex_.AssertHeld();
+#ifndef NDEBUG
+  for (const auto cfd : cfds) {
+    assert(cfd->imm()->NumNotFlushed() != 0);
+    assert(cfd->imm()->IsFlushPending());
+  }
+#endif /* !NDEBUG */
+
+  SequenceNumber earliest_write_conflict_snapshot;
+  std::vector<SequenceNumber> snapshot_seqs =
+      snapshots_.GetAll(&earliest_write_conflict_snapshot);
+
+  auto snapshot_checker = snapshot_checker_.get();
+  if (use_custom_gc_ && snapshot_checker == nullptr) {
+    snapshot_checker = DisableGCSnapshotChecker::Instance();
+  }
+  autovector<Directory*> distinct_output_dirs;
+  std::vector<FlushJob> jobs;
+  int num_cfs = static_cast<int>(cfds.size());
+  // If we reach here, there must be more than one column family in the flush
+  // request. Therefore, we need more superversion_contexts in job_context.
+  // Since superversion_contexts[0] has already created a new superversion, we
+  // create superversions for other elements in superversion_contexts.
+  for (int i = 1; i != num_cfs; ++i) {
+    job_context->superversion_contexts.emplace_back(SuperVersionContext(true));
+  }
+  for (int i = 0; i < num_cfs; ++i) {
+    auto cfd = cfds[i];
+    Directory* data_dir = GetDataDir(cfd, 0U);
+
+    // Add to distinct output directories if eligible. Use linear search. Since
+    // the number of elements in the vector is not large, performance should be
+    // tolerable.
+    bool found = false;
+    for (const auto dir : distinct_output_dirs) {
+      if (dir == data_dir) {
+        found = true;
+        break;
+      }
+    }
+    if (!found) {
+      distinct_output_dirs.emplace_back(data_dir);
+    }
+
+    jobs.emplace_back(
+        dbname_, cfds[i], immutable_db_options_, mutable_cf_options_list[i],
+        &flush_memtable_ids[i], env_options_for_compaction_, versions_.get(),
+        &mutex_, &shutting_down_, snapshot_seqs,
+        earliest_write_conflict_snapshot, snapshot_checker, job_context,
+        log_buffer, directories_.GetDbDir(), data_dir,
+        GetCompressionFlush(*cfd->ioptions(), mutable_cf_options_list[i]),
+        stats_, &event_logger_, mutable_cf_options_list[i].report_bg_io_stats,
+        false /* sync_output_directory */, false /* write_manifest */);
+  }
+
+  autovector<FileMetaData> file_meta;
+  Status s;
+  // TODO (yanqin): parallelize jobs with threads.
+  for (int i = 0; i != num_cfs; ++i) {
+    auto& job = jobs[i];
+    job.PickMemTable();
+    file_meta.emplace_back(FileMetaData());
+
+#ifndef ROCKSDB_LITE
+    // may temporarily unlock and lock the mutex.
+    NotifyOnFlushBegin(cfds[i], &file_meta[i], mutable_cf_options_list[i],
+                       job_context->job_id, job.GetTableProperties());
+#endif /* !ROCKSDB_LITE */
+    if (logfile_number_ > 0 &&
+        versions_->GetColumnFamilySet()->NumberOfColumnFamilies() > 0) {
+      s = SyncClosedLogs(job_context);
```

Comment:
I'm lost here. Are we still holding DB mutex here?

Reply:
Yes.

Reply:
IMO it looks fine since the calls doing I/O (like `SyncClosedLogs`) look like they all release/reacquire the lock.